File: packages/filters-gl/src/dropShadowFilter.ts
import type { GlRenderTarget } from '@flighthq/render-gl';
import { clearGlRenderTarget } from '@flighthq/render-gl';
import type { DropShadowFilter } from '@flighthq/types';
import type { GlRenderState } from '@flighthq/types';

import { applyBoxBlurFilterToGl } from './blurFilter';
import { applyGlBlitOffsetPass, applyGlBlitPass, applyGlTintPass } from './tintShader';

/**
 * Applies a drop shadow filter to `source`, writing the result to `dest`.
 * Compositing order: shadow at offset → source (unless `hideObject` is true).
 *
 * `scratch` must contain three render targets of the same dimensions as `dest`
 * (tint mask, blurred shadow, and the blur's ping-pong temp). The filter
 * allocates nothing itself.
 *
 * `knockout` is not yet supported by the Gl path.
 */
export function applyDropShadowFilterToGl(
  state: GlRenderState,
  source: GlRenderTarget,
  dest: GlRenderTarget,
  scratch: GlRenderTarget[],
  filter: Readonly<Omit<DropShadowFilter, 'type'>>,
): void {
  if (filter.knockout) return;

  const angle = ((filter.angle ?? 45) * Math.PI) / 180;
  const distance = filter.distance ?? 4;
  const dx = Math.cos(angle) * distance;
  const dy = Math.sin(angle) * distance;
  const color = filter.color ?? 0;
  const alpha = filter.alpha ?? 1;
  const strength = filter.strength ?? 1;
  const quality = Math.max(1, Math.round(filter.quality ?? 1));
  const hideObject = filter.hideObject ?? false;

  const tintStrength = Math.min(1, strength);
  const shadowPasses = Math.max(1, Math.floor(strength));

  const [mask, blurred, blurTemp] = scratch;

  applyGlTintPass(state, source, mask, color, alpha, tintStrength);
  applyBoxBlurFilterToGl(state, mask, blurred, blurTemp, {
    blurX: filter.blurX ?? 4,
    blurY: filter.blurY ?? 4,
    passes: quality,
  });

  clearGlRenderTarget(state, dest);
  for (let i = 0; i < shadowPasses; i++) {
    applyGlBlitOffsetPass(state, blurred, dest, dx, dy);
  }

  if (!hideObject) {
    applyGlBlitPass(state, source, dest);
  }
}

File: crates/flighthq-filters-wgpu/src/drop_shadow_filter.rs
//! Drop-shadow wgpu filter pass.
//!
//! Compositing order: shadow at offset → source (unless `hide_object`).
//! `knockout` is not supported by the wgpu path (the call returns early).
//!
//! `scratch` must contain three render targets of the same dimensions as `dest`
//! (`[mask, blurred, blur_temp]`). This function allocates nothing itself.

use flighthq_filters::DropShadowFilter;
use flighthq_render_wgpu::{WgpuRenderState, WgpuRenderTarget};

use crate::blur_filter::apply_box_blur_filter_to_wgpu;
use crate::filter_pass::{WgpuFilterState, clear_wgpu_filter_target};
use crate::tint_shader::{apply_wgpu_blit_offset_pass, apply_wgpu_blit_pass, apply_wgpu_tint_pass};

/// Applies a drop shadow filter to `source`, writing the result to `dest`.
pub fn apply_drop_shadow_filter_to_wgpu(
    state: &mut WgpuRenderState,
    filter_state: &mut WgpuFilterState,
    source: &WgpuRenderTarget,
    dest: &WgpuRenderTarget,
    scratch: &[&WgpuRenderTarget; 3],
    filter: &DropShadowFilter,
) {
    if filter.knockout.unwrap_or(false) {
        return;
    }

    let angle = filter.angle.unwrap_or(45.0).to_radians();
    let distance = filter.distance.unwrap_or(4.0);
    let dx = angle.cos() * distance;
    let dy = angle.sin() * distance;
    let color = filter.color.unwrap_or(0);
    let alpha = filter.alpha.unwrap_or(1.0);
    let strength = filter.strength.unwrap_or(1.0);
    let quality = filter.quality.unwrap_or(1).max(1);
    let hide_object = filter.hide_object.unwrap_or(false);

    // Strength below 1 attenuates the tint; strength above 1 re-blits the shadow that many times.
    let tint_strength = strength.min(1.0);
    let shadow_passes = (strength.floor() as i32).max(1);

    let [mask, blurred, blur_temp] = *scratch;

    apply_wgpu_tint_pass(
        state,
        filter_state,
        source,
        mask,
        color,
        alpha,
        tint_strength,
    );
    apply_box_blur_filter_to_wgpu(
        state,
        filter_state,
        mask,
        blurred,
        blur_temp,
        filter.blur_x.unwrap_or(4.0),
        filter.blur_y.unwrap_or(4.0),
        quality,
    );

    clear_wgpu_filter_target(state, dest);
    for _ in 0..shadow_passes {
        apply_wgpu_blit_offset_pass(state, filter_state, blurred, dest, dx, dy);
    }

    if !hide_object {
        apply_wgpu_blit_pass(state, filter_state, source, dest);
    }
}

#[cfg(test)]
mod tests {}
